Add (frontend-based) end-to-end testing starting point

kschiffer opened the pull request:
#### Summary

References #2605

This PR adds an initial state of (frontend-based) end-to-end testing to our codebase.

![](https://i.imgur.com/Yrl7o2j.gif)

#### Changes

*   Adapt various mage targets to enable end-to-end test runs
    *   Database targets to snapshot, dump and restore the SQL database
    *   Improve target dependencies to enable better caching during CI
    *   Add `endToEnd:*` mage targets to setup and run end-to-end tests in various configurations
*   Add cypress
    *   As well as various related packages, tools and eslint-plugin
*   Add initial cypress tests and utilities
*   Add `End to End Tests` GitHub Action

#### Testing

Manual testing, running and verifying GitHub Actions on my private fork of TTS

##### Regressions

This might break affected `dev/js/jsSdk` mage targets.

#### Notes for Reviewers

<details><summary>Cypress</summary>

So as discussed in #2605, this uses [Cypress](https://www.cypress.io/) as testing engine. I've included a couple of initial tests as well as some utility function that will help us logging in quickly, seeding the database (between runs) and grabbing DOM elements based on the `test-id` attribute.

To run the tests you need to do the following;

*   `mage endToEnd:prepare` only once initially, to build the stack and initialize the DB
*   Either:
    *   Run the stack with `export TTN_LW_IS_DATABASE_URI="postgresql://root@localhost:26257/ttn_lorawan_test?sslmode=disable"`
    *   Run the mage target `mage endToEnd:runTestStack`, which will do the same, but supress the output
*   If you run in development mode, you also need to run `mage js:serve`
*   `mage endToEnd:run`

Cypress offers an interactive mode, in which you can watch and debug the tests as you write them. You can run this via `mage endToEnd:cypressOpen`

The tests will run in a `ttn-lorawan-test` database so that there will be no clashes with the development database. Note that `mage dev:DBErase` will also delete this database.

@bafonins @pgalic96 @mjamescompton (and anyone who wants to write end-to-end tests for that matter), I recommend to read through the [cypress docs](https://docs.cypress.io/guides/overview/why-cypress.html#In-a-nutshell) since we will be making extensive use of these end-to-end tests once this is merged. The docs are very detailed and should give a good grasp of how to write cypress tests.

I also recommend taking note of the [best practices](https://docs.cypress.io/guides/references/best-practices.html) and especially the section about "selecting elements" since we need to avoid selecting elements by arbitrary attributes that might not be persistent. Especially selecting by class is to be avoided, since the rendered classnames include arbitrary hashes. I've also included [`testing-library`](https://github.com/testing-library/cypress-testing-library), which should help us [selecting elements in a way that resembles actual use of the frontend](https://testing-library.com/docs/intro), making the tests more robust and avoiding unnecessary refactors for unrelated code changes as much as possible.

I suggest that we keep track of our decisions of how to write cypress tests in `DEVELOPMENT.md`.

Happy to hear your thoughts, suggestions, etc. Let's try to get a robust setup going.
</details>


<details><summary>Github Action / CI</summary>

You can have a look at the actions [here](https://github.com/kschiffer/lorawan-stack/actions?query=workflow%3A%22End-to-end+Tests%22).

I've spent quite some time to get a good caching mechanism going, which can bring the runtime down to ~4mins in optimal cases. For usual JS changes, it should be around ~6min, which I think is reasonable enough. I've also disabled the caching for pushes to `develop` or `master` to ensure that there are no flakes going through from the caching mechanism.

The action is set up to run for pull requests as well as pushes to `master` or `develop`.

For failed runs, the Github Action will include a screenshot of the failed run as well as the video recordings of all tests. This data can be obtained from the "artifacts" section of the action.

@rvolosatovs there will be quite some conflicts with your recent mage refactors, I guess. I'll try to resolve these once your branches are merged.
</details>


#### Checklist

*   [x] Scope: The referenced issue is addressed, there are no unrelated changes.
*   [x] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md`. The target branch is set to `master` if the changes are fully compatible with existing API, database, configuration and CLI.
*   [x] Documentation: Relevant documentation is added or updated.
*   [x] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
*   [x] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

rvolosatovs commented on .mage/dev.go:
```diff
@@ -85,6 +90,83 @@ func execDockerCompose(args ...string) error {
+	return err
+}
+
+func node() (func(args ...string) error, error) {
```
Is this function really necessary?

kschiffer replied:
Yeah, so instead of redefining it here, we can make the function in `js.go` global.

rvolosatovs replied:
Please just use `sh.Run("node")` instead, there's no reason to have a helper function for this.

rvolosatovs commented on .mage/dev.go:
```diff
@@ -133,6 +215,7 @@ func (Dev) DBRedisCli() error {
 
 // InitStack initializes the Stack.
 func (Dev) InitStack() error {
+	mg.Deps(Dev.DBStart)
```
This should not be a dependency - user might want to run the stack with a locally-running database

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+	return "http://localhost:" + port
+}
+
+func (endToEnd EndToEnd) cypress() (func(args ...string) error, error) {
```
This should depend on `js:Deps` from https://github.com/TheThingsNetwork/lorawan-stack/pull/2751 let's wait for that to be done and merged (@bafonins I currenly working on it)

Other than that:
```suggestion
func (e EndToEnd) cypress() (func(args ...string) error, error) {
```
Also everywhere below

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+
+func (endToEnd EndToEnd) prepareDB() error {
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists("./.cache/sqldump.sql")
+	changed, err := target.Dir("./.cache/sqldump.sql", "./pkg/identityserver/store")
+	if err != nil {
+		return err
+	}
```
Again, dependency on https://github.com/TheThingsNetwork/lorawan-stack/pull/2751
```suggestion
	ok, err := target.Dir("./.cache/sqldump.sql", "./pkg/identityserver/store")
if err != nil {
		return targetError(err)
	}
	// at this point `ok == !dumpExists || changed`, which should be enough.
```

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+func (endToEnd EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.BuildIfNecessary)
```
This is incorrect - `Js.BuildIfNecessary` should have a dependency on `js.Deps` if it needs `Js.Deps`, otherwise this should not be `SerialDeps`

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+	if mg.Verbose() {
+		fmt.Printf("Execute database dump\n")
+	}
+	_ = os.Mkdir(".cache", 0755)
```
Handle errors properly, also use https://golang.org/pkg/os/#MkdirAll instead

kschiffer replied:
Why? This is creating just one folder.

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+	if err != nil {
+		return err
+	}
+
```
nit; remove newline

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+		fmt.Println("Starting stack in end-to-end test configuration")
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	_, err := outputGo("run", "./cmd/ttn-lw-stack", "start")
```
why use `outputGo` if you discard the output?

kschiffer replied:
`execGo` outputs it to stdout. It was an easy way for me to mute the output.

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+	if mg.Verbose() {
+		fmt.Println("Waiting for the stack to be ready...")
+	}
+	for i := 0; i < 100; i++ {
+		resp, _ := http.Get(getTestURL() + "/oauth")
+		if resp != nil && resp.StatusCode == 200 {
+			return nil
+		}
+		time.Sleep(1000 * time.Millisecond)
+	}
```
Not a big fan of this approach. I believe it's possible for `mage` targets to take `context.Context` as parameter, you could then have a
```go
afterCh := time.After(<timeout>)
for {
	resp, err := http.Get(getTestURL() + "/oauth")
	// properly handle errors ...
	if resp.StatusCode == 200 {
		return nil
	}
	select {
		<-case afterCh:
			// return
		<-ctx.Done():
			// return
		default:
			time.Sleep(time.Second)
	}
}
```

kschiffer replied:
If you feel strongly about it, feel free to commit a replacement. Otherwise, this works fine for what I was adding it for.

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// Run starts the end-to-end tests from scratch.
+func (endToEnd EndToEnd) Run() {
```
What's the point of this mage target? Why not just use `RunFrontend`?

kschiffer replied:
The end-to-end test must not all be frontend-based, later on, we might have CLI based end-to-end tests as well. This target setup accounts for that.

rvolosatovs replied:
The keyword is *might*.
We will add it when we need it.

rvolosatovs commented on .mage/end_to_end.go:
```diff
@@ -0,0 +1,163 @@
+}
+
+// RunFrontend starts the frontend end-to-end tests from scratch.
+func (endToEnd EndToEnd) RunFrontend() {
```
What's the point of this mage target? Why not just use `Cypress`?

kschiffer replied:
This is just to keep the target name untangled from cypress, we might choose to use a different library or multiple ones. Bit of a stretch, but it felt cleaner to me that way.

rvolosatovs commented on .mage/js.go:
```diff
@@ -81,29 +81,27 @@ func (js Js) node() (func(args ...string) error, error) {
 }
 
-func (js Js) execFromNodeBin(cmd string, verbose bool) (func(args ...string) error, error) {
+func execFromNodeBin(cmd string, verbose bool, appendArgs ...string) (func(args ...string) error, error) {
```
Please revert change in this function and use `withArguments` from https://github.com/TheThingsNetwork/lorawan-stack/pull/2751

rvolosatovs commented on .mage/js.go:
```diff
@@ -146,7 +144,7 @@ func (js Js) DevDeps() error {
 func (js Js) Deps() error {
 	files, readErr := ioutil.ReadDir("node_modules")
-	changed, targetErr := target.Path("node_modules", "./package.json", "./yarn.lock")
+	changed, targetErr := target.Path("node_modules", "./package.json", "./yarn.lock", "./sdk/js/dist")
```
Please revert this change, this will be handled in https://github.com/TheThingsNetwork/lorawan-stack/pull/2751 - perhaps, you can suggest to apply this there?

kschiffer replied:
Unfortunately forgot to suggest this change there. I'll keep it in this PR.

rvolosatovs commented on .mage/js.go:
```diff
@@ -157,6 +155,7 @@ func (js Js) Deps() error {
 		if err != nil {
 			return err
 		}
+		mg.Deps(JsSDK.Build)
```
Remove

kschiffer replied:
This is necessary for production builds, when the SDK is not linked, as explained in https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442224428

rvolosatovs commented on .mage/js.go:
```diff
@@ -169,7 +168,7 @@ func (js Js) Deps() error {
 // Build runs all necessary commands to build the console bundles and files.
 func (js Js) Build() {
-	mg.SerialDeps(js.Deps, JsSDK.Build, js.BuildDll, js.BuildMain)
+	mg.SerialDeps(js.Deps, js.BuildDll, js.BuildMain)
```
This will be changed in https://github.com/TheThingsNetwork/lorawan-stack/pull/2751

rvolosatovs commented on .mage/js.go:
```diff
@@ -185,6 +184,28 @@ func (js Js) BuildMain() error {
 	return webpack("--config", "config/webpack.config.babel.js")
 }
 
+// BuildIfNecessary builds the frontend bundles but only if source files have
+// been modified.
+func (js Js) BuildIfNecessary() error {
+	isCI := os.Getenv("CI") == "true"
+	publicExists := pathExists("./public")
+	changed, err := target.Dir("./public", "./pkg/webui", "./config/webpack.config.babel.js", "./config/webpack.dll.babel.js", "./sdk/js", "./package.json")
+	if err != nil {
+		return err
+	}
+	if (changed && !isCI) || !publicExists {
+		if mg.Verbose() {
+			fmt.Println("Building frontend files due to changes in source files...")
+		}
+		mg.Deps(js.Build)
+	} else {
+		if mg.Verbose() {
+			fmt.Println("Skipping frontend build due to no source file changes being detected...")
+		}
+	}
+	return nil
+}
+
```
`js:build` *is* `js:buildIfNecessary` - if it's not, it's a bug. Remove
```suggestion
```

kschiffer replied:
From my experience, it's not desirable to apply target checks for the `js.build` as they will always be quite fragile, and result in annoying false positives too often. I really want this command to always execute `js.buildMain`, which is why that target did not check for anything and would always execute, sinxe it's not possible to force execution regardless of targets, as it was in `make`. 

For the CI however, it's good to add these checks to avoid unnecessary builds when the caches are available.

kschiffer replied:
Refactored this to include the CI check in the normal `Js.Build` target.

rvolosatovs commented on .mage/js.go:
```diff
@@ -229,8 +250,13 @@ func (js Js) ServeMain() error {
 
 // Messages extracts the frontend messages via babel.
 func (js Js) Messages() error {
+	isCI := os.Getenv("CI") == "true"
```
This should not be necessary

kschiffer replied:
It shouldn't, but it is unfortunately. When the Github Action runner checks out the repository, it will result in all files being modtimed to the current time. Hence, all mage target checks will return true. The only way to still make use of caching is to use a different target dependency logic for the CI environment.

rvolosatovs commented on .mage/js.go:
```diff
@@ -229,8 +250,13 @@ func (js Js) ServeMain() error {
 // Messages extracts the frontend messages via babel.
 func (js Js) Messages() error {
+	isCI := os.Getenv("CI") == "true"
+	cacheExists := pathExists("./.cache/messages")
```
This should be handled via `target` only, see https://github.com/TheThingsNetwork/lorawan-stack/pull/2751

kschiffer replied:
See https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442275560

rvolosatovs commented on .mage/js_sdk.go:
```diff
@@ -58,8 +58,13 @@ func (k JsSDK) Deps() error {
 
 // Build builds the source files and output into 'dist'.
 func (k JsSDK) Build() error {
+	isCI := os.Getenv("CI") == "true"
```
This should not be necessary

kschiffer replied:
See https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442275560

rvolosatovs commented on .mage/js_sdk.go:
```diff
@@ -191,12 +196,18 @@ func (k JsSDK) DefinitionsClean(context.Context) error {
 
 // Link links the local sdk package via `yarn link` to prevent caching issues.
 func (k JsSDK) Link() error {
+	if nodeEnv := os.Getenv("NODE_ENV"); nodeEnv != "development" {
```
This is incorrect - SDK link is always required, otherwise frontend dependency cannot be found for e.g. `js:lint` and it should work regardless of `NODE_ENV`, cc @bafonins

bafonins replied:
To be more precise, we indeed do not need to link the js sdk in production mode, but we must make sure that the sdk is built before the rest of the frontend. Otherwise, webpack build fails as it cannot find the `ttn-lw` dependency. I took similar approach in https://github.com/TheThingsNetwork/lorawan-stack/pull/2751

rvolosatovs commented on .mage/utils.go:
```diff
@@ -0,0 +1,25 @@
+// Copyright © 2020 The Things Network Foundation, The Things Industries B.V.
```
Remove this file.
Also utils like this should go to `general.go`

rvolosatovs commented on .mage/utils.go:
```diff
@@ -0,0 +1,25 @@
+
+import "os"
+
+func pathExists(filename string) bool {
```
This is not necessary, remove

rvolosatovs commented on .mage/dev.go:
```diff
@@ -85,6 +90,83 @@ func execDockerCompose(args ...string) error {
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	n, err := node()
```
```suggestion
	exe, err := node()
```

bafonins commented on pkg/webui/components/header/index.js:
```diff
@@ -69,7 +69,11 @@ const Header = function({
+            <ProfileDropdown
+              className={style.profileDropdown}
+              userId={user.ids.user_id}
+              data-test="header__profile_dropdown"
```
What is the convention for test attributes?

kschiffer replied:
Yeah, this needs some further thought. I'd suggest to use a [BEM](http://getbem.com/introduction/) inspired approach, mainly to avoid ID clashes. Let's not get too strict about it though.

kschiffer commented on package.json:
```diff
@@ -118,6 +122,9 @@
     "yup": "^0.26.6"
   },
+  "resolutions": {
+    "cypress": "4.6.0"
```
This is to resolve a weird issue with the cypress download in the CI environment: https://github.com/cypress-io/cypress/issues/4840#issuecomment-566303004

htdvisser commented on tools/mage/scripts/restore-db-dump.js:
```diff
@@ -0,0 +1,40 @@
+})
+client.connect()
+const sql = fs.readFileSync('.cache/sqldump.sql', 'utf8')
+client.query('DROP DATABASE ttn_lorawan_test', (err, res) => {
+  client.query('CREATE DATABASE ttn_lorawan_test', (err, res) => {
```
use `dbName` here

bafonins commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,60 @@
+
+// Helper function to quickly seed the database to a fresh state using a
+// previously generated sql dump.
+Cypress.Commands.add('seedDatabase', () => {
```
This is not really `seedDatabase` as you are not seeding any test data. Its fine if we want to add this later, otherwise I would prefer something like `dropDatabase`.

kschiffer replied:
The command will drop the database and fill it with the initial data from the `mage dev:init` target (using the previously generated SQL dump). As such, it is seeding the database, in my understanding.
Most accurate would be `dropAndSeedDatabase` I guess.

kschiffer commented on cypress/integration/user-registration.spec.js:
```diff
@@ -0,0 +1,40 @@
+
+    cy.grab('register__form__user_id')
+      .type('test-user')
+      .should('have.value', 'test-user')
```
It checks whether the foregoing input type simulation has set the value of the input as expected.

kschiffer commented on cypress/integration/login.spec.js:
```diff
@@ -0,0 +1,44 @@
+
+    cy.location('pathname').should('include', '/oauth')
+
+    cy.findAllByText('incorrect password or user ID').should('exist')
```
> I dont think we should check for the message as it is returned by the backend. All we care about here is that the path has no `/console` and the error notification is present on the screen.

I disagree. The goal is to make the tests resemble the actual user experience as closely as possible. In this test case, a user would expect to be shown a notification that informs about the nature of the problem that lead to the submit not being successful. If we only check for the error notification, the test would also pass when there is a notification displaying irrelevant or no information. Generally speaking, the more we match realistic user expectations in our tests, the more confidence we can derive from them.

Also note that since this is an end-to-end test, it is legitimate to make assertions based on backend behavior.

Do you agree?

kschiffer commented on cypress/integration/login.spec.js:
```diff
@@ -0,0 +1,44 @@
+      .should('have.value', 'admin')
+      .type('{enter}')
+
+    cy.grab('overview__welcome_text', { timeout: 10000 }).should('contain.text', 'Welcome')
```
I noticed that the loading time during login can vary quite a lot, especially when running in headless mode. The only way I see to account for that is increasing the timeout. Note that this does not mean that the command will wait for 10 seconds to look for the element. It means that will retry finding the elements for a maximum of 10 seconds (but usually find it sooner than that).

kschiffer commented on cypress/support/commands.js:
```diff
@@ -0,0 +1,60 @@
+})
+
+// Helper function to retrieve elements based on the `data-test` attribute.
+Cypress.Commands.add('grab', (selector, options) => cy.get(`[data-test="${selector}"]`, options))
```
This helper is actually not necessary since we can just use `testing-library`'s `cy.findByTestId()`.

kschiffer commented on cypress/integration/login.spec.js:
```diff
@@ -0,0 +1,44 @@
+// limitations under the License.
+
+describe('Login and OAuth authorization', () => {
+  it('Displays an error on invalid login credentials', () => {
```
I'd like to do that in a separate PR which will also update the coding guidelines and streamlines all test names as a whole.

kschiffer commented:
*   [x] Lets use data-test-id as the test attribute.
*   [x] I would still prefer having documentation in place before merging this pr. Include:
*   [x] Quick description that we use cypress.
*   [ ] We use describe and it to be consistent with already existing tests using jest and avoid using context, specify from mocha. -> will follow up with issue on this
*   [x] How to set e2e tests locally (dashboard access?)
*   [x] Some generic guidelines you used when writing tests. We will define more strict guidelines later.

I've looked into strategies for selecting items again and I think it's best to use the following priority: https://testing-library.com/docs/guide-which-query

Meaning that we only use test id attributes as sort of a last resort and try to use visual or structural cues to select elements otherwise, with which the testing library will help us a lot. This will keep the tests as realistic as possible. I've changed the initial tests accordingly.

kschiffer commented:
@bafonins Thanks for your review. I would like your approval once more, after the changes to the selector strategy I outlined in my last comment. Just to be sure. I've addressed all your remarks as well.

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+          restore-keys: |
+            ${{ runner.os }}-go-
+        if: env.should_cache
+      - name: Apply database dump cache
```
All other workflows call these `Initialize`, not `Apply` - either change all other workflows or conform to the same format.

bafonins commented on tools/mage/js.go:
```diff
@@ -63,20 +63,20 @@ func runYarnV(args ...string) error {
-func (js Js) runEslint(args ...string) error {
-	return js.runYarnCommandV("eslint", append([]string{"--color", "--no-ignore"}, args...)...)
+func (Js) runEslint(args ...string) error {
+	return runYarnCommand("eslint", append([]string{"--color", "--no-ignore"}, args...)...)
```
Dont we want to see `eslint` output when running `js:lint` ?

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+          path: .cache/sqldump.sql
+          key: ${{ runner.os }}-sql-dump-cache-${{ hashFiles('pkg/identityserver/store/**/*.go') }}
+        if: env.should_cache
+      - name: Apply messages and locales cache
```
IDEM

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+          path: |
+            .cache/babel
+            .cache/messages
+            pkg/webui/console/locales/.backend
```
This seems useful for other development builds as well, please copy-paste in all other workflows where we might need this

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+        if: env.should_cache
+      - name: Apply public folder cache
+        uses: actions/cache@v2
+        with:
```
IDEM

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+        if: env.should_cache
+      - name: Apply JS SDK cache
+        uses: actions/cache@v2
+        with:
```
IDEM

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.Build)
```
`js.Build` should and depends on `js.Deps`

kschiffer replied:
I added this because otherwise the `Js.Deps` target would run multiple times in parallel, leading to crashes, due to other dependencies of `Js.Build` depending on `Js.Deps` as well.

rvolosatovs replied:
`mage` should take care of running each dependency at most once - if that does not happen, it's a bug and it should be fixed instead of introducing hacky workarounds.

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (EndToEnd) prepareBuild() {
+	mg.SerialDeps(Js.Deps, Js.Build)
```
```suggestion
	mg.Deps(Js.Build)
```

rvolosatovs commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+      - name: Run test preparations
+        run: tools/bin/mage endToEnd:prepare
+      - name: Run frontend end-to-end tests
+        run: tools/bin/mage endToEnd:startTestStack & tools/bin/mage -v endToEnd:runFrontend
```
So this starts 2 commands in parallel, right?
Please use the
```
|
<cmd>
<cmd>
```
syntax

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+// DBDump performs a database dump to the .cache folder.
+func (EndToEnd) DBDump() error {
+	if mg.Verbose() {
+		fmt.Printf("Execute database dump\n")
```
```suggestion
		fmt.Println("Execute database dump")
```

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+func (EndToEnd) DBRestore() error {
+	mg.Deps(Js.Deps, Dev.DBStart)
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
```
```suggestion
		fmt.Println("Restore database from dump")
```

rvolosatovs commented on DEVELOPMENT.md:
````diff
@@ -714,10 +714,78 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+$ ./mage endToEnd:startTestStack
+```
+
+> Note: This command will suppress the output of executed The Things Stack for LoRaWAN)
````
```suggestion
> Note: This command will suppress the output of executed The Things Stack for LoRaWAN
```

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+		return targetError(err)
+	}
+	if !ok || (isCI && dumpExists) {
+		mg.SerialDeps(EndToEnd.DBRestore)
```
```suggestion
		mg.Deps(EndToEnd.DBRestore)
```

johanstokking commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+      - name: Set up Go 1.14
+        uses: actions/setup-go@v1
+        with:
+          go-version: '1.14'
```
`~1.14`

johanstokking replied:
See other workflows (in `develop`)

rvolosatovs commented on DEVELOPMENT.md:
```diff
@@ -714,10 +714,78 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+
+##### Guding Principle
+
+We follow the [guiding principle of Kent C. Dodds](https://twitter.com/kentcdodds/status/977018512689455106) ([@kentcdodds](https://twitter.com/kentcdodds)) for writing useful tests:
```
I don't think this particular twitter post provides enough detail and I don't think we should make *a* twitter post our guiding principle in general, but it's a nice quote and maybe just put a reference below it?

johanstokking commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+          restore-keys: |
+            ${{ runner.os }}-node-
+        if: env.should_cache
+      - name: Apply go modules cache
```
Lowercase `go`

johanstokking commented on .github/workflows/end-to-end.yml:
```diff
@@ -0,0 +1,110 @@
+        if: env.should_cache
+      - name: Make Mage
+        run: make tools/bin/mage
+      - name: Set up Go 1.14
```
vs `Go`

rvolosatovs commented on config/eslintrc.yaml:
```diff
@@ -158,3 +160,13 @@ rules:
   jsdoc/check-indentation: warn
   jsdoc/require-description-complete-sentence: warn
+
+# Overwrite jest's expect rules for cypress test files, which uses a different
```
does this require `.editorconfig` adjustments as well?

bafonins replied:
No

johanstokking commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+// SQLStart starts the SQL database of the development environment.
+func (Dev) SQLStart() error {
+	if mg.Verbose() {
+		fmt.Printf("Starting SQL databases\n")
```
nit; `Println()` without `\n`

johanstokking commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+// SQLMakeSnapshot stores the current cockroach data folder for later restores.
+func (Dev) SQLMakeSnapshot() error {
+	if mg.Verbose() {
+		fmt.Printf("Making DB snapshot")
```
Missing line ending here?

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
 }
 
+func execDockerComposeWithOutput(filepath string, args ...string) error {
+	output, err := sh.Output("docker-compose", dockerComposeFlags(args...)...)
```
Open the file above and use it as `stdout` in `sh.Exec` call here

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+	if mg.Verbose() {
+		fmt.Printf("Making DB snapshot")
+	}
+	os.RemoveAll(devDataDir + "/cockroach-snap")
```
Do not discard errors, use `filepath.Join`

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+	if mg.Verbose() {
+		fmt.Printf("Restoring DB snapshot")
+	}
+	os.RemoveAll(devDataDir + "/cockroach")
```
Do not discard errors, use `filepath.Join`

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+		fmt.Printf("Making DB snapshot")
+	}
+	os.RemoveAll(devDataDir + "/cockroach-snap")
+	return sh.RunV("cp", "-R", devDataDir+"/cockroach", devDataDir+"/cockroach-snap")
```
IDEM, also prefer using stdlib instead of shelling out to `cp` - different environments may have different implementations and even no `cp` at all (e.g. Windows).

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+		fmt.Printf("Restoring DB snapshot")
+	}
+	os.RemoveAll(devDataDir + "/cockroach")
+	sh.RunV("cp", "-R", devDataDir+"/cockroach-snap", devDataDir+"/cockroach")
```
IDEM

rvolosatovs commented on tools/mage/dev.go:
```diff
@@ -82,6 +87,73 @@ func execDockerCompose(args ...string) error {
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	return sh.Run("node", "./tools/mage/scripts/restore-db-dump.js", "--db", devDatabaseName)
```
IDEM

rvolosatovs commented on tools/mage/general.go:
```diff
@@ -33,3 +34,11 @@ func targetError(err error) error {
 	return nil
 }
+
+func pathExists(filename string) bool {
```
Why not `target.Path` or `target.Dir`?

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442275560

I don't see how that would work. `target.Path` uses modtimes, which cannot be used in the CI caching scenario. It can't work because the checked out files will be modtimed to the current time, meaning that all target checks are returning true, even though a cached value should be used.

Again, if the github action runner runs a target check against a file that has been cached by the action itself, we cannot use the mage target checks, since those would always return `true`, because the modtimes of the cached files will always be older than the freshly checked out source files.

rvolosatovs replied:
https://github.com/magefile/mage/blob/310e198ebd9303cd2c876d96e79de954915f60a7/target/target.go#L27-L30
These checks check existence first. If you care that a path exists only, do not specify dependencies to the call.
This function is redundant.

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -63,20 +63,20 @@ func runYarnV(args ...string) error {
 }
 
-func (Js) runYarnCommand(cmd string, args ...string) error {
+func runYarnCommand(cmd string, args ...string) error {
```
Revert, this is incorrect. These commands must be run in root package, not anywhere else.
If you acknowledge that, you can run `Js.<cmd>`, but there must be a namespace

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -169,6 +171,11 @@ func (js Js) Build() error {
 		fmt.Println("Running Webpack")
 	}
+	isCI := os.Getenv("CI") == "true"
+	publicExists := pathExists("./public")
```
This is a hack, which should not be there.
Please modify the dependencies accordingly instead of this.
Why would this be specific to CI?

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447586088

Please acknowledge.

rvolosatovs commented on tools/mage/js.go:
```diff
@@ -195,7 +202,9 @@ func (js Js) Messages() error {
 	}
-	if !ok {
+	isCI := os.Getenv("CI") == "true"
+	cacheExists := pathExists(filepath.Join(".cache", "messages"))
```
IDEM

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447586088

Please acknowledge.

rvolosatovs commented on tools/mage/js_sdk.go:
```diff
@@ -64,7 +65,9 @@ func (k JsSDK) Build() error {
-	if !ok {
+	isCI := os.Getenv("CI") == "true"
+	distExists := pathExists(filepath.Join("sdk", "js", "dist"))
+	if !ok || (isCI && distExists) {
```
IDEM

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447586088

Please acknowledge.

rvolosatovs commented on tools/mage/go.go:
```diff
@@ -71,6 +71,10 @@ func outputGo(cmd string, args ...string) (string, error) {
 	return buf.String(), nil
 }
 
+func runGoMute(args ...string) error {
```
Instead change `runGo` to have the same behavior as `mage`: https://github.com/magefile/mage/blob/310e198ebd9303cd2c876d96e79de954915f60a7/sh/cmd.go#L64-L69

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// Run starts the end-to-end tests from scratch.
+func (EndToEnd) Run() {
```
This does not do anything, remove

kschiffer replied:
I [commented](https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442262959) before on why I used this.

rvolosatovs replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447592674

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+// RunFrontend starts the frontend end-to-end tests from scratch.
+func (EndToEnd) RunFrontend() {
```
This does not do anything, remove

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+		fmt.Println("Waiting for the stack to be ready...")
+	}
+	for i := 0; i < 100; i++ {
+		resp, _ := http.Get(getTestURL() + "/oauth")
```
Do not discard errors, my earlier comment about this is not addressed.

kschiffer replied:
Contrarily, my [reply to your comment](https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442592462) was not addressed by you.

rvolosatovs replied:
This function will block for more 100 seconds if e.g. `getTestURL` returns wrong URL or host's networking is messed up/different from what we expect and no error will ever be produced.
That is not how quality software is made.
Please either just use the solution I proposed or come up with your own, but this should not happen.

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+		fmt.Println("Starting stack in end-to-end test configuration")
+	}
+	os.Setenv("TTN_LW_IS_DATABASE_URI", databaseURI)
+	return runGoMute("./cmd/ttn-lw-stack", "start")
```
Please add `runGoWith` and use that instead, same as is done in `mage` infrastructure

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+		fmt.Printf("Execute database dump\n")
+	}
+	_ = os.MkdirAll(".cache", 0755)
+	return execDockerComposeWithOutput(filepath.Join(".cache", "sqldump.sql"), "exec", "-T", "cockroach", "./cockroach", "dump", testDatabaseName, "--insecure")
```
break line for readability

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+	if mg.Verbose() {
+		fmt.Printf("Restore database from dump")
+	}
+	return sh.Run("node", filepath.Join("tools", "mage", "scripts", "restore-db-dump.js"), "--db", testDatabaseName)
```
IDEM

kschiffer replied:
> I added this because otherwise the `Js.Deps` target would run multiple times in parallel, leading to crashes, due to other dependencies of `Js.Build` depending on `Js.Deps` as well.

https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447571675

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+	return nil
+}
+
+func (EndToEnd) prepareBuild() {
```
Unnecessary, remove
Just use `mg.Deps(Js.Build)` dependency where necessary

kschiffer replied:
> I added this because otherwise the `Js.Deps` target would run multiple times in parallel, leading to crashes, due to other dependencies of `Js.Build` depending on `Js.Deps` as well.

https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447571675

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists(filepath.Join(".cache", "sqldump.sql"))
+	ok, err := target.Dir(
+		filepath.Join(".cache", "/sqldump.sql"),
```
```suggestion
		filepath.Join(".cache", "sqldump.sql"),
```

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+
+func (EndToEnd) prepareDB() error {
+	isCI := os.Getenv("CI") == "true"
+	dumpExists := pathExists(filepath.Join(".cache", "sqldump.sql"))
```
redundant due to `target.Dir` call

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r447586088

Please acknowledge.

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+}
+
+func (EndToEnd) runCypress(command string, args ...string) error {
+	return runYarnCommand("cypress", append([]string{command, fmt.Sprintf("--config-file=%s", filepath.Join("config", "cypress.json")), "--config", fmt.Sprintf("baseUrl=%s", getTestURL())}, args...)...)
```
IDEM

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+	if nodeEnv == "development" {
+		port = "8080"
+	}
+	return "http://localhost:" + port
```
Use `fmt.Sprintf`

rvolosatovs commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+
+var (
+	testDatabaseName = "ttn_lorawan_test"
+	databaseURI      = "postgresql://root@localhost:26257/" + testDatabaseName + "?sslmode=disable"
```
use `fmt.Sprintf`

bafonins commented on tools/mage/end-to-end.go:
```diff
@@ -0,0 +1,154 @@
+	if mg.Verbose() {
+		fmt.Println("Running end-to-end frontend based tests")
+	}
+	mg.Deps(EndToEnd.Cypress)
```
Why not moving `Cypress` here instead of using `mg.Deps` for chaining target calls?

kschiffer replied:
https://github.com/TheThingsNetwork/lorawan-stack/pull/2764#discussion_r442264114

bafonins commented on cypress/integration/logout.spec.js:
```diff
@@ -0,0 +1,47 @@
+    cy.login()
+    cy.visit('/console')
+    cy.get('header').within(() => {
+      cy.findByText('admin').click()
```
For me it reads like: there is the `profile-dropdown` entry in the `header` and it should display the users id
```suggestion
      cy.findByTestId('profile-dropdown')
      	.should('contain', 'admin')
      	.click()
```
this should possible as `should` returns the same subject it was given

bafonins commented on DEVELOPMENT.md:
````diff
@@ -714,10 +714,78 @@ The message definitions in `pkg/webui/locales` can be used to provide translatio
+To run end-to-end tests, observe the following mage target:
+
+```bash
+$ ./mage endToEnd:run
````
```
$ ./tools/bin/mage -v endToEnd:run                                 
Running target: EndToEnd:Run
Running end-to-end tests
Running dependency: go.thethings.network/lorawan-stack/tools/mage.EndToEnd.RunFrontend
Running end-to-end frontend based tests
Running dependency: go.thethings.network/lorawan-stack/tools/mage.EndToEnd.Cypress
Running dependency: go.thethings.network/lorawan-stack/tools/mage.EndToEnd.WaitUntilReady
Waiting for the stack to be ready...
Running dependency: go.thethings.network/lorawan-stack/tools/mage.Js.Deps
Error: Could not connect to server
```

kschiffer replied:
Additionally, you need to run the stack manually, this is not part of the target.

From the OP:

> To run the tests you need to do the following;
>
> *   `mage endToEnd:prepare` only once initially, to build the stack and initialize the DB
> *   Either:
>    *   Run the stack with `export TTN_LW_IS_DATABASE_URI="postgresql://root@localhost:26257/ttn_lorawan_test?sslmode=disable"`
>    *   Run the mage target `mage endToEnd:runTestStack`, which will do the same, but supress the output
> *   If you run in development mode, you also need to run `mage js:serve`
> *   `mage endToEnd:run`

True though that this info lacks in `DEVELOPMENT.md`

kschiffer commented on cypress/plugins/index.js:
```diff
@@ -0,0 +1,18 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+module.exports = (on, config) => {
```
I added this because cypress will automatically generate this file if it is not there yet.

kschiffer commented:
Closing this since I don't think I will be able to bring this into an agreeable format within a reasonable amount of time. I can't spend more time on this currently and I think there are other people better suited to build this more quickly and thoroughly. If anyone wants to pick this up, feel free to use the branch as foundation/inspiration.

johanstokking commented:
Not so fast. The scope of these tests, end-to-end, mean that there are a lot of concerns to be taken into account. It's perfectly fine to agree to disagree on things – I think all contributors commit to the importance of having end-to-end tests to catch regressions, and that those concerns outweigh concerns about implementation details.

rvolosatovs commented:
I think it's very important to have end-to-end tests, however we should not make our own lives harder by introducing hacks and workarounds in our build process to allow for that - those come with a pretty high maintenance cost and later we end up with the same issue as was with the JS-related mage targets, where everything depends on everything, things break with no apparent reason and issues are very hard to debug/track and fix.
Can't we start with a simple script in github workflows to run an end-to-end test and then properly integrate this into mage tooling at a later point?

bafonins commented:
Will reopen a new PR

johanstokking commented:
@bafonins @kschiffer please delete this branch when no longer necessary
